wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,327 @@
+
+// ConvertAttackType(7) is a hidden attack type with some unique properties. For
+// more information see goo.gl/9k8tn .
+constant ATTACK_TYPE_UNIVERSAL = ConvertAttackType(7)
+constant DAMAGE_TYPE_CHECK_ID = compiletime(ABIL_ID_GEN.next())
+constant CHEAT_DEATH_ID = compiletime(ABIL_ID_GEN.next())
+
+public class Damage
+	integer id
+	unit source
+	unit target
+	bool spell
+	real initial
+	real amount
+
+	/** check this if you want to save target from damage in all cases */
+	protected bool doSaveTarget = false
+	function saveTarget()
+		doSaveTarget = true
+
+	/** Damage instances are automatically destroyed.
+		Use this to protect them if you need that instances somehow. */
+	function saveInstance()
+		wantSave += 1
+
+	/** Releases Damage instance so it can be destroyed. */
+	function releaseInstance()
+		wantSave -= 1
+		if fired and wantSave <= 0
+			destroy this
+
+	protected int wantSave = 0
+	
+	/** stores how much health were added */
+	protected real afterDamage = 0.
+	protected bool killingBlow = false
+	private bool fired = false
+
+	/** applied all modifiers to the damage instance once */
+	function modify()
+		for i in damageModifiers
+			i.run(this)
+	protected function anounce()
+		for i in damageListeners
+			i.run(this)
+
+	/** Returns true if the damage is going to kill or already have killed target */
+	function isKillingBlow() returns bool
+		return killingBlow or (not fired and target.getHP() - .405 < amount)
+		
+	/** Multiplies initial damage with given magnitude */
+	function addMultiplier(real mult)
+		amount += initial * (mult - 1)
+
+	/** Increases damage amount with given amount */
+	function addAmount(real amount)
+		this.amount += amount
+		
+	construct(unit source, unit target, real amount, bool isSpell, int id)
+		this.source = source
+		this.target = target
+		this.initial = amount
+		this.amount = amount
+		spell = isSpell
+		this.id = id
+
+	protected static function processEvent() returns Damage
+		let damage = GetEventDamage()
+		if damage == 0
+			return null
+		let d = new Damage(GetEventDamageSource(), GetTriggerUnit(), (damage < 0 ? -damage : damage), damage < 0, DAMAGE_ID_NONTRIGGERED)
+		d.fired = true
+		return d
+
+	/** applies user-created Damage instance. */
+	function deal()
+		if not fired
+			fired = true
+			if target.isAlive()
+				let life = target.getHP()
+				killingBlow = life - .405 < amount
+				if killingBlow and doSaveTarget
+					amount = target.getHP() - .406
+					killingBlow = false			
+				dealCodeDamage(source, target, amount, killingBlow)
+				anounce()
+			if this.wantSave == 0
+				destroy this
+
+/** Deal an exact amount of damage to `target`, crediting kill experience and
+	gold bounty to `attacker` and not triggering any of DamageDetection listeners. KillingBlow parameter decides whether to kill the target. Isn't supposed to use outside of the system. */
+function dealCodeDamage(unit attacker, unit target, real damage, bool killingBlow)
+	if damage != 0
```

Comment:
What effect does this have for very small epsilon values?

What value does this check have?

Reply:
epsilon values are ignored by '==', not '!=', [real talk](https://www.hiveworkshop.com/threads/real-talk-floats-in-warcraft-3.270579/)

Reply:
What value does this check have?